Add predicates for human reviews (#151)
in-toto/attestation: spec/predicates/human-review-crev.md

# Predicate type: Dependency Reviews (crev)

Type URI: (tentative) https://in-toto.io/attestation/human-review/crev/v0.1

Version: Do we use Crev's versioning? Currently -1? What about their URI?

## Purpose

This attestation type is used to describe the results of human review of
dependency source code. The format is based on the
[crev project](https://github.com/crev-dev/crev).

## Use Cases

As noted above, crev enables social review of popular open source software
dependencies. A crev review includes information such as the thoroughness of
the review, understanding of the source code, and a final rating.

## Prerequisite

Understanding of the

> [in-toto attestation specification](https://github.com/in-toto/attestation).

Reviewer: This isn't really a comment specific to this predicate, but I've seen several predicate specs say this, and I feel like this ought to just be implied. The pre-reqs I think are a lot more interesting and worth noting should be the lesser known, predicate-specific ones.

# Predicate type: Dependency Reviews (crev)

Type URI: (tentative) https://in-toto.io/attestation/human-review/crev/v0.1

> Version: Do we use Crev's versioning? Currently -1? What about their URI?

Reviewer 2: Do we have a resolution for this?  If we don't know maybe just stick with v0.1?

Author: I'll reach out to the crev developers again.

Author: I've added a note for now. The crev developers are open to us reimplementing crev without necessarily being backwards compatible. IMO we should start in a place of compatibility (or close-to-compatibility) and see where we go from there. For now, I've added a note about crev's versioning and updated to using in-toto specific URIs.

## Schema

```json
{
    "_type": "https://in-toto.io/Statement/v1",
    "subject": [{...}],
    "predicateType": "https://in-toto.io/attestation/human-review/crev/v0.1",
    "predicate": {
        "result": "positive|negative",
        "timestamp": "<TIMESTAMP>",
        "reviewer": {
            "idType": "crev",
            "id": "",
            "url": "<URL>"        
        },
        "thoroughness": "high|medium|low",
        "understanding": "high|medium|low",
        "comment": "<STRING>"
    }
}
```

### Parsing Rules

This predicate follows the
[in-toto Attestation Framework's parsing rules](../v1/README.md#parsing-rules).

### Fields

The subject of this predicate type is a specific package and its version in some
ecosystem.

`result`, _enum_, _required_

Specifies if the overall rating of the dependency is `positive` or `negative`.

`reviewer` _object_, _required_

Identifies the reviewer. This has some meaning for crev's trust proliferation
aspects, but the identity of the reviewer can also be mapped based on in-toto's
functionary handling. `idType` is used to determine the contents of `reviewer`.
The `url` is a reference to the reviewer's crev-proofs repository.

> `timestamp` _Timestamp_, _required_

Reviewer 2: nit: instead of calling this 'timestamp' which leaves what time it refers to ambiguous, can it be named review_time or something more descriptive?

See https://github.com/in-toto/attestation/issues/46 for some discussion

# Predicate type: Dependency Reviews (crev)

Type URI: (tentative) https://in-toto.io/attestation/human-review/crev/v0.1

Note: At the time of writing, the crev project uses a separate versioning
scheme, with the current version being -1. In future, this predicate may adopt
a URI associated with that project.

Version: v0.1.0

## Purpose

This attestation type is used to describe the results of human review of
dependency source code. The format is based on the
[crev project](https://github.com/crev-dev/crev).

## Use Cases

As noted above, crev enables social review of popular open source software
dependencies. A crev review includes information such as the thoroughness of
the review, understanding of the source code, and a final rating.

## Model

Most modern software have external dependencies. Dependency review is the
process of reviewing and verifying the source code of a particular
dependency, and can be performed by one or more of several actors in the supply
chain. The developer importing a new dependency can perform the review or a
dedicated security team can be tasked with it.

> ## Schema

Reviewer: To my more general comment about which Crev spec is authoritative, I suggest indicating here which format this predicate follows. As indicated on the main project repo, I'm assuming the spec specified in [cargo-crev](https://github.com/crev-dev/cargo-crev/blob/master/crev-lib/rc/doc/editing-code-review.md) is the authoritative one.

```json
{
    "_type": "https://in-toto.io/Statement/v1",
    "subject": [{...}],
    "predicateType": "https://in-toto.io/attestation/human-review/crev/v0.1",
    "predicate": {
        "result": "positive|negative",
        "reviewer": {
            "idType": "crev",
            "id": "<ID>",
            "url": "<URL>"        
        },
        "reviewTime": "<TIMESTAMP>",
        "thoroughness": "high|medium|low",
        "understanding": "high|medium|low",
        "comment": "<STRING>"
    }
}
```

### Parsing Rules

This predicate follows the
[in-toto Attestation Framework's parsing rules](../v1/README.md#parsing-rules).

### Fields

The subject of this predicate type is a specific package and its version in some
ecosystem.

`result`, _enum_, _required_

> Specifies if the overall rating of the dependency is `positive` or `negative`.

Reviewer: We should probably support the full set of values for the "rating" field of the [Code Review Proofs](https://github.com/crev-dev/cargo-crev/blob/master/crev-lib/rc/doc/editing-code-review.md).

`reviewer` _object_, _required_

Identifies the reviewer. This has some meaning for crev's trust proliferation
aspects, but the identity of the reviewer can also be mapped based on in-toto's
functionary handling. `idType` is used to determine the contents of `reviewer`.

> The `url` is a reference to the reviewer's crev-proofs repository.

Reviewer: A question and a suggestion. Q: What are possible `idType` values, and who determines these? Rec: I'd add in here that this field is intended to correspond to the `from` field in the Code Review Proofs format when the `idType` matches the ID for Crev.

Author: There's a bit of a disconnect here, because the crev [code review](https://github.com/crev-dev/cargo-crev/blob/master/crev-lib/rc/doc/editing-code-review.md) type includes `from` but missing from the [package review](https://github.com/crev-dev/cargo-crev/blob/master/crev-lib/rc/doc/editing-package-review.md) type. Same for `date`. I suspect `date` may be omitted as a package-wide review is meant as an ongoing document with updates as new versions / advisories emerge. On the other hand, I'm not sure about the `from` field. I wonder if in the in-toto context we want to drop it and lean entirely on the signature. That takes us to a broader conversation about how in-toto supports crev, though.

`reviewTime` _Timestamp_, _required_

> Indicates time of review creation. `timestamp` in the original crev
> specification.

Reviewer: This field seems to have been renamed.
```suggestion
Indicates time of review creation. `date` in the original crev
specification.
```

`thoroughness` _enum_, _required_

Describes how thorough the reviewer was. Must be set to one of `low`, `medium`,
or `high`.

`understanding` _enum_, _required_

Describes the reviewer's understanding of the dependency code. Must be set to

> one of `low`, `medium`, or `high`.

Reviewer: As I mentioned in my comment above, I'd suggest we support using all possible values for these fields according to the [cargo-crev](https://github.com/crev-dev/cargo-crev/blob/master/crev-lib/rc/doc/editing-code-review.md#data-fields) data fields spec.

`comment` _string_, _optional_

Optional field with any other comments a reviewer may have about the

> dependency.

Reviewer: Which consumers, if any, of this predicate need to be able to parse/care about the contents of the `comment` field? This would be helpful info.

This attestation type is used to describe the results of human review of
dependency source code. The format is based on the
[crev project](https://github.com/crev-dev/crev), specifically the
[package review proof](https://github.com/crev-dev/cargo-crev/blob/master/crev-lib/rc/doc/editing-package-review.md)
defined in the [cargo-crev](https://github.com/crev-dev/cargo-crev)
implementation.

## Use Cases

crev enables social review of popular open source software dependencies. A crev
review includes information such as the thoroughness of the review,

> understanding of the source code, and a final rating. The ratings for these

Reviewer: Being very nitpicky here, though I think this conveys the expectations more clearly.
```suggestion
understanding of a package's source code, and a final rating. The ratings for these
```

fields are self-identified by each individual reviewer.

## Model

Most modern software have external dependencies. Dependency review is the

> process of reviewing and verifying the source code of a particular dependency,

Reviewer: Nit:
```suggestion
process of reviewing and verifying the source code of a particular open-source package,
```

functionary handling. `idType` is used to determine the contents of `reviewer`.
The `url` is a reference to the reviewer's crev-proofs repository.

`date` _Timestamp_, _required_

Indicates time of review creation. `timestamp` in the original crev
specification.

`thoroughness` _enum_, _required_

Describes how thorough the reviewer was. Must be set to one of `low`, `medium`,

> `high`, or `none`.

Reviewer: Semantically, what does a `none` value tell a consumer of the attestation? Could this be interpreted as an "endorsement" of a particular package?
